hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -2778,6 +2777,15 @@ private void prepareForSelection(SqmPath<?> joinedPath) {
 			}
 			else if ( joinedPath instanceof SqmTreatedRoot<?, ?> ) {
 				navigablePath = ( (SqmTreatedRoot<?, ?>) joinedPath ).getWrappedPath().getNavigablePath();
+
+				final TableGroup tableGroupForTreatedRoot = getTableGroupForTreatedRoot(
+						navigablePath,
+						fromClauseIndex
+				);
+				if ( tableGroupForTreatedRoot != null ) {
+					fromClauseIndex.register( joinedPath, tableGroupForTreatedRoot );
+					return;
+				}
```

Comment:
I think you should add the following to line 2750 in prepareReusablePath

```
if ( sqmPath instanceof SqmTreatedPath<?, ?> ) {
	fromClauseIndex.register( (SqmPath<?>) sqmPath, tableGroup );
}
```

right after `implicitJoinChecker.accept( newTableGroup );`